dask/dataframe/rolling.py

```diff
@@ -91,31 +146,174 @@ def map_overlap(func, df, before, after, *args, **kwargs):
+
+    kwargs3 = {}
+    simple = True
+    for k, v in kwargs.items():
+        v = normalize_arg(v)
+        v, collections = unpack_collections(v)
+        dependencies.extend(collections)
+        kwargs3[k] = v
+        if collections:
+            simple = False
+
+    if has_keyword(func, "partition_info"):
+        partition_info = {
+            (i,): {"number": i, "division": division}
+            for i, division in enumerate(divisions[:-1])
+        }
+
+        args2.insert(0, BlockwiseDepDict(partition_info))
+        orig_func = func
+
+        def func(partition_info, *args, **kwargs):
+            return orig_func(*args, **kwargs, partition_info=partition_info)
+
+    if enforce_metadata:
+        dsk = partitionwise_graph(
+            apply_and_enforce,
+            name,
+            func,
+            before,
+            after,
+            *args2,
+            dependencies=dependencies,
+            _func=overlap_chunk,
+            _meta=meta,
+            **kwargs3,
+        )
     else:
-        meta = _emulate(func, df, *args, **kwargs)
-    meta = make_meta(meta, index=df._meta.index, parent_meta=df._meta)
+        kwargs4 = kwargs if simple else kwargs3
+        dsk = partitionwise_graph(
+            overlap_chunk,
+            name,
+            func,
+            before,
+            after,
+            *args2,
+            **kwargs4,
+            dependencies=dependencies,
+        )
+
+    graph = HighLevelGraph.from_collections(name, dsk, dependencies=dependencies)
+    return new_dd_object(graph, name, meta, divisions)
 
-    name = f"{func_name}-{token}"
-    name_a = "overlap-prepend-" + tokenize(df, before)
-    name_b = "overlap-append-" + tokenize(df, after)
-    df_name = df._name
 
-    dsk = {}
+def _get_nexts_partitions(df, after, dsk):
+    df_name = df._name
 
     timedelta_partition_message = (
         "Partition size is less than specified window. "
         "Try using ``df.repartition`` to increase the partition size"
     )
 
+    name_b = "overlap-append-" + tokenize(df, after)
+    if after and isinstance(after, Integral):
+        nexts = []
+        for i in range(1, df.npartitions):
+            key = (name_b, i)
+            dsk[key] = (M.head, (df_name, i), after)
+            nexts.append(key)
+        nexts.append(None)
+    elif isinstance(after, datetime.timedelta):
+        # TODO: Do we have a use-case for this? Pandas doesn't allow negative rolling windows
+        deltas = pd.Series(df.divisions).diff().iloc[1:-1]
+        if (after > deltas).any():
+            raise ValueError(timedelta_partition_message)
+
+        nexts = []
+        for i in range(1, df.npartitions):
+            key = (name_b, i)
+            dsk[key] = (_head_timedelta, (df_name, i - 0), (df_name, i), after)
+            nexts.append(key)
+        nexts.append(None)
+    else:
+        nexts = [None] * df.npartitions
+    return nexts
+
+
+def _get_previous_partitions(df, before, dsk):
```

Comment:
`_get_previous_partitions` and `_get_nexts_partitions` both mutate `dsk`, which isn't best practice for Dask.

We're curious, why have you refactored these into separate functions?

Reply:
I refactored these into separate functions because I need to call this code for every _Frame object in args. Before my change, the code was inlined and executed once (only one _Frame argument were supported).
Moreover, because they are now independent from map_overlap, we can add specific unit test on it. 
I can change the Api to not mutate dsk

Reply:
API changed to not mutate dsk